### PR TITLE
Script to generate the translations base file

### PR DESCRIPTION
**Pull request by dbkr**
Also replaces the translations base file with the output of the
script.

Just adding this as script right now: if it works out OK, we can look at running it automatically at some point in the build pipeline (and using it for the strings in vector-web too).

The JSON file does have different keys and is not just reordered, here's a more helpful summary of what's changed:

```
925 old strings, 905 new strings

34 missing from new:
and
restore
Thursday
And %(count)s more...
Sunday
you must be a
%(weekDayName)s, %(monthName)s %(day)s %(fullYear)s
demote
times
tag direct chat
or
to browse the directory
Monday
Friday
had
Failed to register as guest:
Uploading %(filename)s and %(count)s others|zero
Matrix Apps
favourite
Wednesday
rejected
%(names)s and %(count)s others are typing
joined and left
%(weekDayName)s, %(monthName)s %(day)s
left and rejoined
Never send encrypted messages to unverified devices in this room
Flair
Set
Tuesday
to start a chat with someone
joined
to make a room or
Saturday
left

14 extra in new:
Unnamed room
World readable
No pinned messages.
%(names)s and %(count)s others are typing|other
%(names)s and %(count)s others are typing|one
'%(groupId)s' is not a valid group ID
And %(count)s more...|other
Unpin Message
Guests can join
Invalid group ID
And %(count)s more...|one
No rooms to show
Loading...
Jump to message
```

I think all the removed ones are unused (the days of the week, some fragments of strings, old labs feature labels, etc) and adds some missed ones.

I think the only non-obvious removal is `Uploading %(filename)s and %(count)s others|zero` as I only included the 'one' and 'others' plurals, but it's nonsensical to have a message for uploading zero files as we don't show the bar in that situation.

### CODE REVIEW

**Comment by pafcu** [comment]
This seems to be hardcoded to look under src/ and .js files only, and will fail with riot-web which also has translatable content in res/home.html.

File: scripts/gen-i18n.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+const walk = require('walk');
+
+const flowParser = require('flow-parser');
+const estreeWalker = require('estree-walker');
+
+const TRANSLATIONS_FUNCS = ['_t', '_td', '_tJsx'];
+
+// A selection of plural variants to put in the base file: other langauges
+// with more types of plural will have more, but they will just be in the file
```
Comment (lukebarnard1) [inline scripts/gen-i18n.js]: Maybe clarify that we only need to worry about English plurals here because the base file is English

**Comment by pafcu** [comment]
It might also be worth using a word other than "base file" as that has a different meaning in Weblate, so there is some risk of confusion.

**Comment by dbkr** [comment]
Luke raises a good point (IRL) that we don't have all the info we need in the source for the plurals because we only have one string (generally the 'other' string) and the 'one' string is only in the en_EN translation file, so we can't generate the en_EN file from the source alone. Back to the drawing board. :(

I guess the only things we could do here are either use the old en_EN file as input or keep the en_EN strings in a separate file for translations with counts.

**Comment by lukebarnard1** [comment]
>This seems to be hardcoded to look under src/ and .js files only, and will fail with riot-web which also has translatable content in res/home.html.

**Comment by lukebarnard1** [comment]
oops :sweat_smile:

**Comment by pafcu** [comment]
> Luke raises a good point (IRL) that we don't have all the info we need in the source for the plurals because we only have one string (generally the 'other' string) and the 'one' string is only in the en_EN translation file, so we can't generate the en_EN file from the source alone.

Yes, that's why I just add and remove keys in my script, rather than re-generate the whole thing.

File: scripts/gen-i18n.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
+/**
+ * Regenerates the translations base file by walking the source tree and
+ * parsing each file with flow-parser. Emits a JSON file with the
+ * translatable strings mapped to themselves in the order they appeared
+ * in the files and grouped by the file they appeared in.
```
[the remark this answers is not in the document]
Comment (dbkr) [reply scripts/gen-i18n.js]: I do quite like that translations from the same file all live together

File: scripts/gen-i18n.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    estreeWalker.walk(tree, {
+        enter: function(node, parent) {
+            //console.log(node);
+            if (
+                node.type == 'CallExpression' &&
+                TRANSLATIONS_FUNCS.includes(node.callee.name)
+            ) {
+                const tKey = getTKey(node.arguments[0]);
+                if (tKey === null) return;
```
[the remark this answers is not in the document]
Comment (dbkr) [reply scripts/gen-i18n.js]: This does happen whenever we pass non-literals to `_t` (ie. wherever we've had to use `_td` to compensate, so just logging when this is null wouldn't be enough (I'll comment this)

**Comment by dbkr** [comment]
Right, I have made this take a translations file as input and copy and other plural variants in that file into the output. It's currently set to use the output file as the input too, but we could also have the input file as a separate file with the extra en_EN plural variants in it (it's the same code).

I've not committed the new en_EN.json file yet as I'll do that just before merging (and pull in the latest from weblate too to minimise conflicts)

**Comment by pafcu** [comment]
I would still try to make it easier to look for other files than .js files in src/, because e.g. riot-web does have translatable strings in other places that this code will miss.

**Comment by dbkr** [comment]
Well, flow-parser only parses JS, so it's only going to work for JS files anyway.

**Comment by pafcu** [comment]
True. That means that it's not possible to use this tool for riot-web as you can't even manually add the strings without them getting automatically removed

**Comment by dbkr** [comment]
Yep. The strings in the homepage are just done with a simple regex anyway so that would be easy to fix, although I'd like to revisit how we do i18n on the homepage anyway as it only works with the bundled homepage (since it uses the translation strings bundled with riot).

**Comment by lukebarnard1** [comment]
Then I say we de-scope that, we certainly should revisit homepages in general (especially considering the new possibilities with groups)

**Pull request merged**